### PR TITLE
Completely rewrite the symbiotic2cs.py script

### LINKED ISSUE #1: TODO: refresh the symbiotic2cs.py
There have been some updates to symbiotic's output messages. Check the $(SUBJ)  script and update it accordingly.

### PR DESCRIPTION
**Pull request by jamartis**
Completely new version of the symbiotic2cs.py script. Hopefully a bit more readable. 
Missing features: 
~~- properly check and update the file path names so the error messages do not point to the symbiotic files~~
~~- parse the info and non deterministic values sections~~

The output messages slightly differ compared to the previous script, benchmark will most likely fail due to this.

### CODE REVIEW

File: symbiotic2cs.py [hunk 1]
```diff
@@ -17,12 +17,23 @@ def __init__(self,argv = ""):
...
     def fix_file(self):
         'Check whether the file is a file from the input package or file from some symbiotic library. If is the laterthis function tries to find a proper file name/line number from the stackinfo'
-        'TODO'
+        if self.stack != "":
+            for l in str.splitlines(self.stack):
+                match = re.search("\s*note: call stack: function (.*) at: (.*):([0123456789]*)\s*",l)
+                if match == None:
+                    continue
+                else:
+                    if re.search("/opt/symbiotic", match.group(2)):
```
Comment (jamartis) [inline symbiotic2cs.py:28]: This regexp might need some further tweaks

**Review by lzaoral — CHANGES_REQUESTED** [review]
Thanks for the rewrite @jamartis! I've tried it with aufover-benchmark and, unfortunately, all converted results seem to be unparsable by `csgrep`. For example `predator-regre-test-0002` results in the following converted output:
```
Error: SYMBIOTIC_WARNING:
test-0002.c:12: memory error: out of bound pointer
test-0002.c
test-0002.c:12
test-0002.c:12: note: Additional Info: address: 0:0
test-0002.c:12: note: Additional Info: pointing to: none
<Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [4 times 0x0] (i32: 0)

Error: SYMBIOTIC_WARNING:
test-0002.c:10: memory error: out of bound pointer
test-0002.c
test-0002.c:10
test-0002.c:10: note: Additional Info: address: 0:12370169555311111083
test-0002.c:10: note: Additional Info: pointing to: none
<Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [0x1|3 times 0x0] (i32: 1)
```
and `csgrep` outputs:
```
output-con.txt:2: parse error: wrong token: T_UNKNOWN (expected T_EVENT)
output-con.txt:3: parse error: wrong token: T_UNKNOWN (expected T_EVENT)
output-con.txt:4: parse error: wrong token: T_UNKNOWN (expected T_EVENT)
Error: SYMBIOTIC_WARNING:
test-0002.c:12: note: Additional Info: address: 0:0
test-0002.c:12: note: Additional Info: pointing to: none
<Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [4 times 0x0] (i32: 0)
output-con.txt:10: parse error: wrong token: T_UNKNOWN (expected T_EVENT)
output-con.txt:11: parse error: wrong token: T_UNKNOWN (expected T_EVENT)
output-con.txt:12: parse error: wrong token: T_UNKNOWN (expected T_EVENT)

Error: SYMBIOTIC_WARNING:
test-0002.c:10: note: Additional Info: address: 0:12370169555311111083
test-0002.c:10: note: Additional Info: pointing to: none
<Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [0x1|3 times 0x0] (i32: 1)
```

It looks like something is missing after both the second and third occurrences of `test-0002.c` substring and there seems to be a missing newline on the last one as well. Are these bugs or am I just doing something wrong?

EDIT: be a bit more precise

**Review by lzaoral — APPROVED** [review]

**Review by lzaoral — CHANGES_REQUESTED** [review]
Thanks for the fixup! If I have just a single suggestion:

Lines like these contain the location where the non-deterministic choice was made so it should not be unknown (patched symbiotic that makes them absolute is building in copr right now): 
```
<Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0006.c:12:18: len 4 bytes, [4 times 0x0] (i32: 0)
```

By the way, the output that I have in my previous comment is from the benchmark, so the benchmarks would need some additional changes to not produce this garbled output and also need a prefix of the absolute paths stripped for portability.

EDIT: copr link: https://copr.fedorainfracloud.org/coprs/g/aufover/symbiotic/build/2832685/

**Comment by jamartis** [comment]
Thanks for the review. 

For the unknown:unknown part: I am actually not sure what to put here to make it the least confusing:
a) The same values as in the previous lines, which might differ from the value in the actual note message?
b) the values from the note message itself (test-0006.c:12:18 in this case): If this differs from a) just in the line number, will people actually notice it?
c) The current state (unknown) - this IMHO forces the user to read the actual note a notice that the file/line nr. might be different
d) something different?

edit: un-MARKDOWN this message

**Comment by lzaoral** [comment]
I'd go for b) as that is the location where the non-deterministic choice was made. @kdudka what is your opinion on this?

**Comment by kdudka** [comment]
I think we should rewrite:
```
<Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [4 times 0x0] (i32: 0)
```
like this:
```
test-0002.c:9:9: note: Non-deterministic values: __VERIFIER_nondet_int: len 4 bytes, [4 times 0x0] (i32: 0)
```

That is the format that gcc-compatible compilers use by default (and what csdiff recognizes).

**Review by lzaoral — CHANGES_REQUESTED** [review]
I've just found another problem. All lines from call stack repeat the location info twice:
```
/home/lukas/aufover-benchmark/workdir/single-c/mem-basic-realloc/0029-realloc-invalid-ptr-freed-2/symbiotic/0029-test.c:10: note: call stack: function main () at: /home/luka
s/aufover-benchmark/workdir/single-c/mem-basic-realloc/0029-realloc-invalid-ptr-freed-2/symbiotic/0029-test.c:10
```

**Comment by jamartis** [comment]
> I think we should rewrite:
> 
> ```
> <Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [4 times 0x0] (i32: 0)
> ```
> 
> like this:
> 
> ```
> test-0002.c:9:9: note: Non-deterministic values: __VERIFIER_nondet_int: len 4 bytes, [4 times 0x0] (i32: 0)
> ```
> 
> That is the format that gcc-compatible compilers use by default (and what csdiff recognizes).

Ok, I'll rewrite it like that.






> I've just found another problem. All lines from call stack repeat the location info twice:
> 
> ```
> /home/lukas/aufover-benchmark/workdir/single-c/mem-basic-realloc/0029-realloc-invalid-ptr-freed-2/symbiotic/0029-test.c:10: note: call stack: function main () at: /home/luka
> s/aufover-benchmark/workdir/single-c/mem-basic-realloc/0029-realloc-invalid-ptr-freed-2/symbiotic/0029-test.c:10
> ```

Not exactly, the first location is based on the location of the error, the second location is based on the function call: 
```
#include <stdlib.h>

int bar ()
{
        int *a = NULL;
        *a = 10;
        return *a;
}

int foo ()
{
        return bar();
}


int main ()
{
        return foo();

}
```

```foo.c:6: error: memory error: out of bound pointer
foo.c:6: note: call stack: function bar () at: foo.c:6
foo.c:6: note: call stack: function foo () at: foo.c:12
foo.c:6: note: call stack: function main () at: foo.c:18
foo.c:6: note: Additional Info: address: 0:0
foo.c:6: note: Additional Info: pointing to: none
```
(using an old version of symbiotic, thus the relative paths)

**Comment by kdudka** [comment]
> ```
> foo.c:6: note: call stack: function bar () at: foo.c:6
> foo.c:6: note: call stack: function foo () at: foo.c:12
> foo.c:6: note: call stack: function main () at: foo.c:18
> foo.c:6: note: Additional Info: address: 0:0
> foo.c:6: note: Additional Info: pointing to: none
> ```

This is not optimal from user's point of view because it renders vim's quickfix unusable.  We really need to move the specific locations to left.  If one loads the report with `:cfile ...` in vim, it should jump over the source code according to call stack on each `:cnext`.  The problem is, of course, not specific to vim.  One would hit the same problem with Eclipse, Visual Studio and the like.

**Comment by lzaoral** [comment]
I think that the prefix of such lines should be the location in the call stack as both `cbmc` and `divine` and even the old `symbiotic` formatter did that:

```
Error: DIVINE_WARNING:
test.c: scope_hint: In function 'bar':
test.c:6: error: null pointer dereference: [global*]
test.c:6: note: memory error in userspace
/opt/divine/include/dios/sys/fault.hpp:119: note: void __dios::FaultBase::handler<__dios::Context>(_VM_Fault, _VM_Frame*, void (*)())
test.c:6: note: bar
test.c:12: note: foo
test.c:18: note: main
/opt/divine/include/dios/libc/sys/start.cpp:91: note: __dios_start

Error: CBMC_WARNING:
test.c: scope_hint: In function ‘bar’:
test.c:6: error: pointer_dereference : dereference failure: pointer NULL in *a
test.c:16: note: from_entry_point_function:function-call:main
test.c:18: note: from_main:function-call:foo
test.c:12: note: from_foo:function-call:bar
test.c:5: note: bar:5:pointer:        a = ((signed int *)NULL)
test.c:6: note: dereference failure: pointer NULL in *a

Error: SYMBIOTIC_WARNING:
/home/lukas/tmp/test.c:6: error: memory error: out of bound pointer
/home/lukas/tmp/test.c:6: note: call stack: bar ()
/home/lukas/tmp/test.c:12: note: call stack: foo ()
/home/lukas/tmp/test.c:18: note: call stack: main ()
```

**Review by lzaoral — CHANGES_REQUESTED** [review]
If Symbiotic crashes, the converter should inform us so that we can investigate the issue and potentially report it to the upstream. The following output gets completely silenced:

```
8.0.0-pre-llvm-12.0.1-symbiotic:13ca9347-dg:0b1ada38-sbt-slicer:b0864a5b-sbt-instrumentation:648fabcc-klee:6ecf91e2
INFO: Looking for invalid dereferences, invalid free, memory leaks, etc.
INFO: Starting instrumentation
RESULT: ERROR (Instrumenting the code failed)
```

**Comment by kdudka** [comment]
> If Symbiotic crashes, the converter should inform us so that we can investigate the issue and potentially report it to the upstream. The following output gets completely silenced:
> 
> ```
> 8.0.0-pre-llvm-12.0.1-symbiotic:13ca9347-dg:0b1ada38-sbt-slicer:b0864a5b-sbt-instrumentation:648fabcc-klee:6ecf91e2
> INFO: Looking for invalid dereferences, invalid free, memory leaks, etc.
> INFO: Starting instrumentation
> RESULT: ERROR (Instrumenting the code failed)
> ```

We could convert it to an `internal error` event to be compatible with gcc: https://github.com/csutils/csdiff/blob/e2b387254dc10218866ba04add3c0d3ade376672/src/parser-common.hh#L27

If there is no location info, it could be something like this:
```
Error: SYMBIOTIC_WARNING:
symbiotic: internal error: Instrumenting the code failed
```

EDIT: I forgot the `Error: SYMBIOTIC_WARNING:` headline.

**Comment by kdudka** [comment]
Of course, if Symbiotic provided at least path to the source (or bitcode) file that triggered the internal error, it would be much more useful.

**Comment by lzaoral** [comment]
That info is always present if you use `--debug=all` as the contents of `argv` are the first thing printed to the output. Also note that the error explanation in `()` may be sometimes missing so in that case something like `unspecified error` would be sufficient.

**Comment by lzaoral** [comment]
This is just a nice-to-have, but it would be great if the script could parse files passed as arguments instead of `stdin` if there are any.

**Comment by lzaoral** [comment]
Thanks a lot! I've tested the fix-up and both `stdin` and files passed as arguments can now be parsed.

**Comment by jamartis** [comment]
I've added a check for the `RESULT: ERROR` line. However, I've tested this only with an input file with missing main function. I am not really sure how consistent is Symbiotic with its internal error reporting so the last commit might still need some further tweaks.

**Comment by lzaoral** [comment]
LGTM! I've modified `symbiotic` to print-out source location info instead of the corresponding LLVM IR instruction and this PR will need some additional tweaks (may contain bugs though): https://copr.fedorainfracloud.org/coprs/jamartis/symbiotic/build/2874845/ You can use the `membasic-free` tests for some model outputs or generally any test with 

Example (`predator/predator-regre-test-0010`):
```
 --- Error trace ---

Error: memory error: memory not cleaned up
File: /home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c
Line: 77
assembly.ll line: 38
Stack:
        #000000038 in main () at /home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:77
Info:
        address: 48:0
        pointing to: object at Segment: [48] of size 8
                MO37[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 66:0
        pointing to: object at Segment: [66] of size 8
                MO55[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 72:0
        pointing to: object at Segment: [72] of size 8
                MO61[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 78:0
        pointing to: object at Segment: [78] of size 8
                MO67[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 84:0
        pointing to: object at Segment: [84] of size 8
                MO73[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 90:0
        pointing to: object at Segment: [90] of size 8
                MO79[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 96:0
        pointing to: object at Segment: [96] of size 8
                MO85[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 102:0
        pointing to: object at Segment: [102] of size 8
                MO91[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 108:0
        pointing to: object at Segment: [108] of size 8
                MO97[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
        address: 114:0
        pointing to: object at Segment: [114] of size 8
                MO103[8] allocated at create_item():/home/lukas/aufover-benchmark/workdir/single-c/predator/predator-regre-test-0010/symbiotic/test-0010.c:13:call
```

Note that the there are some corner cases in which the location info is not emitted by `clang` and there is nothing I can do about it. One such case is are global string constants:
```c
extern void free(void*);
int main(void) { free("this"); }
```

**Comment by lzaoral** [comment]
And looks like I've found the first bug in my patch, the variable name is sometimes missing and sometimes present. I'll look into it tomorrow.

EDIT: Oh, I was running `symbiotic` with `--gen-ll` option to automatically generate LLVM IR in human readable form and it additionally passes `-fno-discard-value-names` to `clang` to preserve the variable names, so use `--gen-ll` when testing the outputs.

**Comment by lzaoral** [comment]
Alright, `-fno-discard-value-names` is not a sufficient way to preserve variable names as every instruction in LLVM IR in the same function must have a unique name. Therefore, shadowed variables will be named differently.

**Review by lzaoral — APPROVED** [review]
